Public profile page

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

This PR adds an initial version of the public profile page, that we'll use to show several information about a participant. For now:

* Avatar.
* Username.
* Nickname.
* Contact link.
* ~Personal description.~ Not yet

At later iterations, we'll add:

* Officialization status and information about it (#2336).
* The user's list of activities (#1139) (has privacy implications).

#### :pushpin: Related Issues
- Fixes #1955.

#### :clipboard: Subtasks
- [x] Minimal profile page.
- ~[ ] Enhace profile page with user description.~
- [x] Change all current dummy avatar links to point to the public profile page.

### :camera: Screenshots (optional)
![w0k95b6c kassulke wunsch and parisian](https://user-images.githubusercontent.com/2887858/34180261-c8a2bd80-e4ec-11e7-89cf-a7b8d0211f60.png)

#### :ghost: GIF
![lions](https://user-images.githubusercontent.com/2887858/34071027-7fc05d7a-e24e-11e7-930d-1f9920ec9db5.gif)

deivid-rodriguez commented:
Any preferences on URLs? I went with the simple `/profile/:nickname`, but there's other possibilities such as the geekier `/~:nickname`.

Feel free to tell me to stop bikeshedding and move on! :stuck_out_tongue:

mrcasals commented:
I've always felt `/~:nickname` was hard to write on non-English keyboards, in case someone wants to manually type the URL... I'd go with `/profile/:nickname`, but I don't have a strong opinion on this

deivid-rodriguez commented:
I think that's a good reason, agreed!

xabier commented:
Why not follow the Twitter startandard? What is the function of the ‘:‘ symbol on the URL?

mrcasals commented:
@xabier The `:` is used to differentiate what is static from what might change: `/profile/:nickname` translates to `/public/xabier`, for example :smile:

deivid-rodriguez commented:
@xabier We could follow twitter's standard `/:username`, but it complicates things a bit since it needs more careful handling of route matching. We'd need to keep a blacklist of usernames, such as `admin`, `system`, `api`, `account` and every other "root route", and prevent those from being registered to avoid route colissions.

xabier commented:
With @mrcasals 's explanation is good! adding `profile` before nickname is good for me. When mentioning twitter standard I was just worried about removing `:` symbol from the url (which is solved now :)

deivid-rodriguez commented:
I'm building this on top of #2335, but I'm actually adding further refactorings to the code in that PR. Since merging of that PR is getting delayed, maybe I should supersed that PR with this one? :thinking:

deivid-rodriguez commented:
Actually forget about it, let's keep PR's as small and focused as possible.

deivid-rodriguez commented:
This is ready.

I decided to not add the description field here, because my design skills are terrible, so it takes me a lot of time to make things look relatively decent. The page will be properly designed later anyways so I prefer to unblock this so other dependent PRs can move on (officialization and #2401). I can come back later to the description field when maybe they'll be some UI cards for it.

This PR sits on top of #2360, and contains some changes I didn't expect to make here (which I can split off if you want me to):

#### JS dependency upgrade

I was touching the graphQL API and I was getting several problems:
  * The server would keep crashing because it would automatically query comments for a `Decidim::Resources::Resource` which does not exist in the development environment (it's only for tests). No idea why that was happening but I couldn't get rid of it.
  * I wouldn't be able to decrease apollo's polling interval to get faster feedback about my changes. I googled and I read that it was due to a bug in `react-apollo` 1.x.

So I upgraded JS dependencies and it fixed both things... :man_shrugging:

#### GraphQL author interface improvements & simplification

The interface to display authors in comments was a bit weird because it would query a `isUser` field and do type checking in the client side. This seems to defeat the point of defining an interface, so I removed it. Also, I added a new "badge" field, that I expect to use in the next PR to display the "officialization badge".

xabier commented:
Nice one @deivid-rodriguez , this PR has been referenced as blocking #2401 but this should be ready now! we are slowly but nicely connecting the dots for social-network-type boost of Decidim :smile:

deivid-rodriguez commented on decidim-comments/app/frontend/application/apollo_client.ts:
```diff
@@ -1,16 +1,20 @@
+  link: httpLink,
+  cache: new InMemoryCache(),
 });
 
```
All this stuff is me following upgrade guides for `apollo-client` 2.x... But I don't really know what I'm doing :laughing:

deivid-rodriguez commented on decidim-comments/app/frontend/comments/add_comment_form.component.tsx:
```diff
@@ -419,26 +417,28 @@ const AddCommentFormWithMutation = graphql<addCommentMutation, AddCommentFormPro
-                commentable: {
-                  ...prev.commentable,
-                  comments,
+            if (prev) {
```
Ts started failing after the upgrade saying that this variable could unexpectedly be `nil` sometimes. So I checked for that.

deivid-rodriguez commented on decidim-comments/app/frontend/comments/comment.component.tsx:
```diff
@@ -209,7 +268,7 @@ class Comment extends React.Component<CommentProps, CommentState> {
     const { session, comment, votable, rootCommentable, orderBy } = this.props;
 
-    if (votable) {
+    if (session && votable) {
```
Another "variable could expectedly be nil" error (don't remember the exact copy but it can be reproduced by removing the check :))

deivid-rodriguez commented on decidim-comments/app/frontend/comments/down_vote_button.component.tsx:
```diff
@@ -55,62 +55,60 @@ const downVoteMutation = require("../mutations/down_vote.mutation.graphql");
-        mutate({
-          variables: {
-            id: ownProps.comment.id,
+  props: ({ ownProps, mutate }: { ownProps: DownVoteButtonProps, mutate: MutationFunc<DownVoteMutation> }) => ({
```
This error was very weird (to me, anyways). Since it was strangely only happening to "down voting", I fixed it by changing this to look exactly like "up voting" does.

deivid-rodriguez commented on decidim-core/app/uploaders/decidim/avatar_uploader.rb:
```diff
@@ -8,11 +8,11 @@ class AvatarUploader < ImageUploader
 
     version :big do
-      process resize_and_pad: [500, 500]
+      process resize_to_fit: [40, 40]
```
I needed to change this to `resize_to_fit`, otherwise the avatar was distorted. This version was previously unused, now I'm using it for the profile page, and optimized the size.

josepjaume replied:
I don't really know why we didn't spot this bafore, but we should be using double resolution sizes (probably 80x80 in this case) to comply with retina displays.

deivid-rodriguez commented on decidim-core/app/uploaders/decidim/avatar_uploader.rb:
```diff
@@ -8,11 +8,11 @@ class AvatarUploader < ImageUploader
 
     version :thumb do
-      process resize_and_pad: [100, 100]
+      process resize_to_fit: [20, 20]
```
Same here with image distorsion, I wonder why we didn't see it before. I can dig in further to confirm.

josepjaume replied:
Same resolution doubling here

deivid-rodriguez commented on decidim-core/lib/decidim/core/api/user_group_type.rb:
```diff
@@ -22,14 +22,16 @@ module Decidim
       resolve ->(obj, _args, _ctx) { UserGroupPresenter.new(obj).avatar_url }
     end
 
-    field :isVerified, !types.Boolean, "Whether the user group is verified or not", property: :verified?
```
Also removed this field from the API since the `badge` field seemed enough. I could add it back with a better name (to account for officialization, like `hasBadge`), but I think think it's worth it just to make a single check in the client side more readable (`hasBadge === true` instead of `badge === ""`). Let me know anyways.

deivid-rodriguez commented:
Added some inline comments to help out reviewers!

oriolgual commented on decidim-core/app/controllers/decidim/profiles_controller.rb:
```diff
@@ -0,0 +1,20 @@
+
+    helper_method :user
+
+    def show; end
```
Can we skip action declaration or do you prefer being explicit?

deivid-rodriguez replied:
I prefer being explicit but would that actually work anyways? :thinking:

oriolgual replied:
Yup, if Rails fins a `show`file it would render it

deivid-rodriguez replied:
Ha, interesting! So as long as I define the route and the template, I don't need the controller action, right? Wow :)

oriolgual replied:
Yup!

oriolgual commented on decidim-core/app/controllers/decidim/profiles_controller.rb:
```diff
@@ -0,0 +1,20 @@
+    private
+
+    def user
+      @user ||= Decidim::User.find_by!(nickname: params[:nickname])
```
Should we scope this to the current organization?

deivid-rodriguez replied:
Absolutely! The nickname is not even unique accross organizations! I'll update!

josepjaume commented:
@deivid-rodriguez do you think we could move the comments upgrade to a separate PR? It looks like it might generate bugs and we should be cautious.

deivid-rodriguez commented:
You mean the JS upgrade? Absolutely!

deivid-rodriguez commented:
Changes made, waiting for confirmation on the changes to be extracted.

oriolgual commented:
OK to me, waiting on @josepjaume

deivid-rodriguez commented:
Thanks for your feedback @oriolgual, so glad you're back to working on decidim! :)

josepjaume commented:
If it isn't too much trouble, I'd be great to have them on a different PR so I can ask @beagleknight for a quick check.

deivid-rodriguez commented:
Yup! I'll do it right now!

josepjaume commented:
Thanks!! ❤️

oriolgual commented:
@deivid-rodriguez #2402 merged! Could you rebase this one please? ❤️

deivid-rodriguez commented:
R.I.P. (Rebase in Progress) :smiley:

deivid-rodriguez commented:
@oriolgual Rebased and fixed [a bug](https://github.com/decidim/decidim/pull/2391/commits/3a7fea5bdde62b277851ddbcc9af66f5866d23f7) that I noticed in the process!

deivid-rodriguez commented on decidim-comments/app/frontend/comments/comment.component.tsx:
```diff
@@ -157,20 +157,42 @@ class Comment extends React.Component<CommentProps, CommentState> {
+   */
+  private _renderActiveAuthor() {
+    const { comment: { author } } = this.props;
+
```
The message of this commit is now incorrect because half of it was included in #2402 (to fix codeclimate there). Worth rebasing?

deivid-rodriguez replied:
Not doing it, we could edit it right before merge & squash directly.
